lambda/translcomprehension.ml

```diff
@@ -2,304 +2,406 @@ open Lambda
 open Typedtree
 open Asttypes
 
-type comp_block =
-  | Unguarded of lambda * array_kind
-  | Guarded of lambda * array_kind * int
+let int n = Lconst (Const_base (Const_int n))
 
-type arrays =
-  | Array_of_elements
-  | Array_of_arrays of arrays
-  | Array_of_filtered_arrays of arrays
+type binding =
+  { let_kind : let_kind;
+    value_kind : value_kind;
+    var : Ident.t;
+    init : lambda }
 
-let int n = Lconst (Const_base (Const_int n))
+let binding let_kind value_kind var init =
+  {let_kind; value_kind; var; init}
 
-let empty_arr ~loc=
-  Lprim( (Pmakearray(Pgenarray, Immutable)), [] ,loc)
+let gen_binding {let_kind; value_kind; var; init} body =
+  Llet(let_kind, value_kind, var, init, body)
 
-let make_array size init ~loc=
-  let prim_make_arr =
+let gen_bindings bindings body =
+  List.fold_right gen_binding bindings body
+
+(* Translate a clause into some initialising bindings, a variable
+   that will be bound to the number of iterations in the clause by
+   those bindings, and lambda code that performs the iterations. *)
+let transl_arr_clause ~transl_exp ~scopes ~loc clause body =
+  let len_var = Ident.create_local "len_var" in
+  let bindings, for_ =
+    match clause with
+    | In (pat , e2) ->
+        let in_var = Ident.create_local "in_var" in
+        let in_kind = Typeopt.array_kind e2 in
+        let in_binding =
+          binding Strict Pgenval in_var (transl_exp ~scopes e2)
+        in
+        let len_binding =
+          let init = Lprim( (Parraylength(in_kind)), [Lvar(in_var)], loc) in
+          binding Alias Pintval len_var init
+        in
+        let index = Ident.create_local "index" in
+        let for_ =
+          Lfor(index, (int 0), Lprim(Psubint, [Lvar(len_var); int 1], loc) , Upto,
+            Matching.for_let ~scopes pat.pat_loc
+              (Lprim(Parrayrefu(in_kind),
+                     [Lvar(in_var); Lvar(index)], loc)) pat body)
+        in
+        [in_binding; len_binding], for_
+    | From_to(id, _, e2, e3, dir) ->
+        let from_var = Ident.create_local "from" in
+        let from_binding =
+          binding Strict Pintval from_var (transl_exp ~scopes e2)
+        in
+        let to_var = Ident.create_local "to" in
+        let to_binding =
+          binding Strict Pintval to_var (transl_exp ~scopes e3)
+        in
+        let low, high =
+          match dir with
+          | Upto -> Lvar from_var, Lvar to_var
+          | Downto -> Lvar to_var, Lvar from_var
+        in
+        let len_binding =
+          let init =
+            Lprim(Psubint, [Lprim(Paddint, [high; int 1], loc); low], loc)
+          in
+          binding Alias Pintval len_var init
```

Comment:
This should be strict and not Alias since len_var has at least 2 occurrences.

Reply:
This is fine because the `init` expression is pure.